torch/distributed/device_mesh.py

```diff
@@ -1038,6 +1038,86 @@ def _flatten(
+        ) -> "DeviceMesh":
+            """
+            Returns a DeviceMesh by unflatten the current DeviceMesh.
+
+            This api can be used to unflatten a N-D DeviceMesh into N-1+len(mesh_sizes)-D meshes or submeshes.
+            The dim is the dimension to be unflattened which can be either a string or an integer.
+            
+            The mesh_sizes is a tuple which specifies the shape of the mesh unflatten into for the given dim.
+            The mesh_dim_names is a list of strings which specifies the names of the dimensions of the mesh unflatten into.
+            Its length must match the length of mesh_sizes.
+            
+            For example, if we have a 1D mesh DeviceMesh([0, 1, 2, 3, 4, 5, 6, 7], mesh_dim_names=("world")),
+            calling mesh_1d._unflatten(0, (2, 2, 4), ["dp", "pp", "tp"]) will create a 3D mesh
+            DeviceMesh([[[0, 1], [2, 3]], [[4, 5], [6, 7]]], mesh_dim_names=("dp", "cp", "tp")).
+            
+            After calling the unflatten, to access the unflattened dimension in mesh_1d, one can use the
+            existing slicing method to obtain the unflattened mesh through calling mesh_1d["dp"].
+            """
+            if isinstance(dim, int) and dim >= self.ndim:
+                raise ValueError(
+                    f"dim {dim} specified in `_unflatten` is out of range {self.ndim}"
+                )
+            elif isinstance(dim, str) and dim in not_none(self.mesh_dim_names):
+                raise ValueError(
+                    f"dim {dim} specified in `_unflatten` is not in {self.mesh_dim_names}"
+                )
+
+            if len(mesh_sizes) != len(mesh_dim_names):
+                raise RuntimeError(
+                    "mesh_dim_names must have same length as mesh_sizes in _unflatten!"
+                )
+
+            if not self._backend:
+                raise NotImplementedError(
+                    "flatten a device mesh without backend initialized is not supported!"
+                )
+
+            strides = [1]
+            for i in range(len(mesh_sizes) - 1):
+                strides.append(strides[-1] * mesh_sizes[i])
+            strides.reverse()
+
+            if isinstance(dim, str):
+                dim = not_none(self.mesh_dim_names).index(dim)
+            
+            unflatten_layout = _Layout(tuple(zip(mesh_sizes, strides)))
+            unflattened_layout, unflattened_dim_names = [], []
+            idxs = [i + dim for i in range(len(mesh_sizes))]
+            for i in range(self.ndim):
+                if i == dim:
+                    unflattened_layout.extend(self._layouts[i].composition(unflatten_layout))
+                    unflattened_dim_names.extend(mesh_dim_names)
+                else:
+                    unflattened_layout.append(self._layouts[i])
+                    unflattened_dim_names.append(not_none(self.mesh_dim_names)[i]) 
+
+            global_override = _mesh_resources.mesh_dim_group_options.get(
```

Comment:
what is this override?

Reply:
it is the pg option change brought in https://github.com/pytorch/pytorch/pull/159371. This part needs more love as well, but in general it is not related to bookkeeping.